src/utils/Cypher.ts

```diff
@@ -26,19 +32,79 @@ export default class Cypher {
     return this.configuration;
   }
 
-  public static encrypt(data: string): string {
+  public static async detectConfigurationKey(tenantID: string): Promise<void> {
+
+    const configCryptoKey: string = Configuration.getCryptoConfig().key;
+    const keySettings = await SettingStorage.getCryptoKeySettings(tenantID);
+    let cryptoKeySetting: CryptoKeySetting;
+
+    // Check if Crypto Key settings exists
+    if (keySettings) {
+      // Detect new config Cypher key
+      if (keySettings.cryptoKey.newKey && keySettings.cryptoKey.newKey !== configCryptoKey) {
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.newKey
+        } as CryptoKeySetting;
+      } else if (!keySettings.cryptoKey.newKey && keySettings.cryptoKey.oldKey !== configCryptoKey) {
+        // If no newKey exist, detect oldKey change
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.oldKey
+        } as CryptoKeySetting;
+      }
+      // If key change detected, prepare KeySettings change
+      if (cryptoKeySetting) {
+        await SettingStorage.deleteSetting(tenantID, keySettings.id);
+        cryptoKeySetting.newKey = configCryptoKey;
+
+        // Key migration of senitive data
+        this.cryptoKey = cryptoKeySetting;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoKeySetting = {
+        oldKey: configCryptoKey
+      } as CryptoKeySetting;
+    }
+
+    if (cryptoKeySetting) {
+      const keySettingToSave = {
+        identifier: TenantComponents.CRYPTO_KEY,
+        type: KeySettingsType.CRYPTO_KEY,
+        cryptoKey: cryptoKeySetting
+      } as KeySettings;
+      await SettingStorage.saveCryptoKeySettings(tenantID, keySettingToSave);
+    }
+  }
```

Comment:
That method looks really overcomplicated to me: 
+ ciphering bootstrap: only one key per tenant and initial sensitive data cyphering will use it;
+ key updated: two keys key, formerKey and no migration done: sensitive data cyphering will still use the formerKey until the migration is done successfully. No more key update is allowed since the migration is done. So the crypto settings data structure is more like :

```
{
	key: string;
	formerKey?: string;
	migrationDone?: boolean;
}
```

With that kind of data structure, that function becomes a test on a boolean attribute. It then just has to be properly updated at key change and ciphering migration done.

Reply:
I saw an email about that topic but do not have the time to answer before today. 
It seems that the code now reflect the design recommandation. 
Do you still have issue with it?

Reply:
I tried to simplify the method. Is it okay now? I still have to integrate the migrationDone part